pkg/api/requestcontext.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+)
+
+type RequestContextMapper interface {
+	Get(req *http.Request) (Context, bool)
+	Update(req *http.Request, context Context) error
+}
+
+type requestContextMap struct {
+	contexts map[*http.Request]Context
+	lock     sync.Mutex
+}
+
+func NewRequestContextMapper() RequestContextMapper {
+	return &requestContextMap{
+		contexts: make(map[*http.Request]Context),
+	}
+}
+
+func (c *requestContextMap) Get(req *http.Request) (Context, bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	context, ok := c.contexts[req]
+	return context, ok
+}
+
+func (c *requestContextMap) Update(req *http.Request, context Context) error {
```

Comment:
I think so.

----- Original Message -----
> > +}
> > +
> > +func NewRequestContextMapper() RequestContextMapper {
> > +	return &requestContextMap{
> > +		contexts: make(map[*http.Request]Context),
> > +	}
> > +}
> > +
> > +func (c *requestContextMap) Get(req *http.Request) (Context, bool) {
> > +	c.lock.Lock()
> > +	defer c.lock.Unlock()
> > +	context, ok := c.contexts[req]
> > +	return context, ok
> > +}
> > +
> > +func (c *requestContextMap) Update(req *http.Request, context Context)
> > error {
> 
> And yet, to allow the auth layer to update the context for the request to one
> with the user value set, something has to let us update the map (but only if
> it exists, so we can be sure it will get cleaned up?)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4352/files#r24539620